clang/utils/TableGen/ClangAttrEmitter.cpp

```diff
@@ -1161,6 +1161,36 @@ namespace {
     void writeHasChildren(raw_ostream &OS) const override { OS << "true"; }
   };
 
+  class IntegerExprArgument : public ExprArgument {
+  private:
+    int UndefValue;
+
+  public:
+    IntegerExprArgument(const Record &Arg, StringRef Attr)
+        : ExprArgument(Arg, Attr), UndefValue(Arg.getValueAsInt("UndefValue")) {
+    }
+
+    void writeAccessors(raw_ostream &OS) const override {
+      OS << "  " << getType() << " get" << getUpperName() << "() const {\n";
+      OS << "    return " << getLowerName() << ";\n";
+      OS << "  }\n";
+      OS << "private:\n";
+      OS << "  int " << getLowerName() << "Value = " << UndefValue << ";\n";
+      OS << "public:\n";
+      OS << "  int get" << getUpperName() << "Evaluated(ASTContext &Ctx) {\n";
+      OS << "    if (" << getLowerName() << "Value != " << UndefValue << ")\n";
+      OS << "      return " << getLowerName() << "Value;\n";
+      OS << "    llvm::APSInt Val(32);\n";
+      OS << "    bool Succeedded = " << getLowerName()
+         << "->isIntegerConstantExpr(Val, Ctx);\n";
+      OS << "    assert(Succeedded && \"expression must be constant "
+            "integer\");\n";
+      OS << "    " << getLowerName() << "Value = Val.getSExtValue();\n";
+      OS << "    return " << getLowerName() << "Value;\n";
+      OS << "  }";
+    }
+  };
+
```

Comment:
Wow, I haven't seen such stuff before.
+ @erichkeane , could you please review and comment on approach?

Reply:
I don't think this is the right approach, we shouldn't be adding helper functions to the tablegen types like this.  Typically we use these only when we need to represent new argument types.

While I get the wish to have a 'cached' value, evaluating the expression only happens a handful of times.  Additionally, the effort to replace the constant-evaluator includes a 'memory' of evaluated values. When this 

If we REALLY need it short term, we generally put a helper function into AdditionalMembers.

Reply:
> While I get the wish to have a 'cached' value, evaluating the expression only happens a handful of times.

If I understand correctly, it can happen a lot of times and actually depends on how code is written. Basically we compare `IntelReqdSubGroupSizeAttr` object which attached to a kernel function with *each* `IntelReqdSubGroupSizeAttr` object attached to any callee (or callee of callee - we analyze the whole call graph) of the kernel. Some applications just put this attribute into a macro and attach it to almost every single function, which means significant amount of evaluations

> Additionally, the effort to replace the constant-evaluator includes a 'memory' of evaluated values. When this

Seems like unfinished sentence 

> If we REALLY need it short term, we generally put a helper function into AdditionalMembers.

This is what I did originally, but then @MrSidims said that he needs similar thing for several other attributes and I tried to make something generic and re-usable here

> I don't think this is the right approach, we shouldn't be adding helper functions to the tablegen types like this. Typically we use these only when we need to represent new argument types.

I'm open to other suggestions, right now I see two other ways:
- Just a helper function somewhere to evaluate `Expr *`, something like [this](https://github.com/intel/llvm/pull/1905/commits/8a35cf0a2a8643599bc3a10d65bedb28963d7b80#diff-c88a590aca6d5f8e92dd1f8b6ec05a49L312). No caching, no modifications to tablegen
- Common class for our attributes with Expr arguments, which declares `AdditionalMembers` - we will still be able to re-use it for other attributes, but we will have to get rid of "named" attribute arguments to do so ([this comment](https://github.com/intel/llvm/pull/1905#discussion_r442322546) from @MrSidims)

Reply:
> 
> 
> > While I get the wish to have a 'cached' value, evaluating the expression only happens a handful of times.
> 
> If I understand correctly, it can happen a lot of times and actually depends on how code is written. Basically we compare `IntelReqdSubGroupSizeAttr` object which attached to a kernel function with _each_ `IntelReqdSubGroupSizeAttr` object attached to any callee (or callee of callee - we analyze the whole call graph) of the kernel. Some applications just put this attribute into a macro and attach it to almost every single function, which means significant amount of evaluations
> 

Potentially, but it also depends on the cost of the expression.  Is anyone using massive constexpr functions to generate these?  if it is just a Constant integer or something (or a sizeof/simple expression), the cost is still minimal.

> > Additionally, the effort to replace the constant-evaluator includes a 'memory' of evaluated values. When this
> 
> Seems like unfinished sentence

Yeah, I meant to say, when that happens, this isn't going to be necessary.
> 
> > If we REALLY need it short term, we generally put a helper function into AdditionalMembers.
> 
> This is what I did originally, but then @MrSidims said that he needs similar thing for several other attributes and I tried to make something generic and re-usable here
> 
> > I don't think this is the right approach, we shouldn't be adding helper functions to the tablegen types like this. Typically we use these only when we need to represent new argument types.
> 
> I'm open to other suggestions, right now I see two other ways:
> 
>     * Just a helper function somewhere to evaluate `Expr *`, something like [this](https://github.com/intel/llvm/pull/1905/commits/8a35cf0a2a8643599bc3a10d65bedb28963d7b80#diff-c88a590aca6d5f8e92dd1f8b6ec05a49L312). No caching, no modifications to tablegen

Something like that seems to be fine to me.  If we want to cache an Expr* to int-result relationship, we can do so there.  We do something similar (using a densemap) for record size vs recorddecl, so a similar solution here seems right.

> 
>     * Common class for our attributes with Expr arguments, which declares `AdditionalMembers` - we will still be able to re-use it for other attributes, but we will have to get rid of "named" attribute arguments to do so ([this comment](https://github.com/intel/llvm/pull/1905#discussion_r442322546) from @MrSidims)

I don't think we want to remove named arguments.  That doesn't seem like a good idea.  And I don't really see how it solves this problem.

Reply:
> Potentially, but it also depends on the cost of the expression. Is anyone using massive constexpr functions to generate these? if it is just a Constant integer or something (or a sizeof/simple expression), the cost is still minimal.

I'm not sure about massive `constexpr` stuff, most likely this is just a template argument or something like this

Thanks for the feedback. I will then probably revert everything back to a single helper function for integer evaluation (and check if something like this already exists and we could re-use that)

Reply:
Lets fix the bug first and then we can discuss, what refactoring we want/can do. 

> I don't think we want to remove named arguments. That doesn't seem like a good idea. And I don't really see how it solves this problem.

For example we have two attributes `reqd_sub_group_size` and `num_simd_work_items`. Both of them are single argument of positive integer attributes. To handle them we have two functions: `addReqdSubGorupSizeAttr` (which calls `getReqdSubGroupSize`) and `addNumSimdWorkItemsAttr` (which calls `getNumSimd`). Since both of the attributes shares the same semantic, we can replace these two functions with `addSingleArgPositiveIntAttr` (which shall can `getValue`).

Reply:
Well, actually, `addSomeAttr` not a candidate for re-usability here, instead `instantiateSYCLOneArgFunctionAttr` is.

Reply:
> Lets fix the bug first and then we can discuss, what refactoring we want/can do.

Reverted all my experiments in 3bee3df561c2cd9c3f3873946986d19764c24906